x/globalfee/feechecker/feechecker.go

```diff
@@ -0,0 +1,147 @@
+package feechecker
+
+import (
+	"errors"
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/bandprotocol/chain/v2/x/globalfee"
+	"github.com/bandprotocol/chain/v2/x/globalfee/types"
+	oraclekeeper "github.com/bandprotocol/chain/v2/x/oracle/keeper"
+	oracletypes "github.com/bandprotocol/chain/v2/x/oracle/types"
+)
+
+type FeeChecker struct {
+	OracleKeeper    *oraclekeeper.Keeper
+	GlobalMinFee    globalfee.ParamSource
+	StakingSubspace paramtypes.Subspace
+}
+
+func NewFeeChecker(
+	oracleKeeper *oraclekeeper.Keeper,
+	globalfeeSubspace, stakingSubspace paramtypes.Subspace,
+) FeeChecker {
+	if !globalfeeSubspace.HasKeyTable() {
+		panic("global fee paramspace was not set up via module")
+	}
+
+	if !stakingSubspace.HasKeyTable() {
+		panic("staking paramspace was not set up via module")
+	}
+
+	return FeeChecker{
+		OracleKeeper:    oracleKeeper,
+		GlobalMinFee:    globalfeeSubspace,
+		StakingSubspace: stakingSubspace,
+	}
+}
+
+func (fc FeeChecker) CheckTxFeeWithMinGasPrices(
+	ctx sdk.Context,
+	tx sdk.Tx,
+) (sdk.Coins, int64, error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return nil, 0, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	feeCoins := feeTx.GetFee()
+	gas := feeTx.GetGas()
+
+	// Ensure that the provided fees meet minimum-gas-prices and globalFees,
+	// if this is a CheckTx. This is only for local mempool purposes, and thus
+	// is only ran on check tx.
+	if ctx.IsCheckTx() {
+		// Check if this is a valid report transaction, we allow gas price to be zero and set priority to the highest
+		isValidReportTx := fc.CheckReportTx(ctx, tx)
+		if isValidReportTx {
+			return sdk.Coins{}, int64(math.MaxInt64), nil
+		}
+
+		requiredFees := getMinGasPrice(ctx, feeTx)
+		requiredGlobalFees, err := fc.GetGlobalFee(ctx, feeTx)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		allFees := CombinedFeeRequirement(requiredGlobalFees, requiredFees)
```

Comment:
Why not combine gas price first then calculate required fee to compare with transaction fee later?

Reply:
updated